Userland/Libraries/LibGUI/FileSystemModel.cpp

```diff
@@ -260,10 +260,28 @@ FileSystemModel::FileSystemModel(String root_path, Mode mode)
         dbgln("Event at \"{}\" on Node {}: {}", node.full_path(), &node, event);
 
         // FIXME: Your time is coming, un-granular updates.
-        node.has_traversed = false;
-        node.mode = 0;
-        node.children.clear();
-        node.reify_if_needed();
+        auto refresh_node = [](Node& node) {
+            node.has_traversed = false;
+            node.mode = 0;
+            node.children.clear();
+            node.reify_if_needed();
+        };
+
+        if (event.type == Core::FileWatcherEvent::Type::Deleted) {
+            auto canonical_event_path = LexicalPath::canonicalized_path(event.event_path);
+            if (m_root_path.starts_with(canonical_event_path)) {
+                // Deleted directory contains our root, so navigate to our nearest parent.
+                auto new_path = LexicalPath(m_root_path).dirname();
+                while (!Core::File::is_directory(new_path))
+                    new_path = LexicalPath(new_path).dirname();
```

Comment:
it's pretty sad to reparse this every time, perhaps we could add `LexicalPath::parent()` and use that here?

Reply:
Agreed, that makes sense. Right now though, someone else is changing LexicalPath's internals, so I've just added a FIXME for this for now. (Because any changes I make will need to be redone later.)

Reply:
That works :P

Reply:
Now using LexicalPath::parent(). :^)